odoo/fields.py

```diff
@@ -1195,9 +1195,21 @@ def __get__(self, record, owner):
                 value = env.cache.get(record, self)
 
             elif self.store and record._origin and not (self.compute and self.readonly):
-                # new record with origin: fetch from origin
-                value = self.convert_to_cache(record._origin[self.name], record, validate=False)
-                env.cache.set(record, self, value)
+                # new record with origin: fetch from origin, and assign the
+                # records to prefetch in cache (which is necessary for
+                # relational fields to "map" prefetching ids to their value)
+                recs = record._in_cache_without(self)
+                try:
+                    for rec in recs:
+                        if rec._origin:
+                            value = self.convert_to_cache(rec._origin[self.name], rec, validate=False)
+                            env.cache.set(rec, self, value)
```

Comment:
```suggestion
                    for rec in recs:
                        if (origin_rec := rec._origin):
                            value = self.convert_to_cache(origin_rec[self.name], rec, validate=False)
                            env.cache.set(rec, self, value)
```
Maybe to avoid recreating rec._origin twice?

Reply:
Odoo 16 requires Python 3.7 minimum, and the walrus operator has been introduced in Python 3.8.
I will adapt this in forward-ports.